RBAC system with custom roles integrated with AWX

AlanCoding opened the pull request:
This rewrites the roles system inside of AWX and moves them here, to this library, so that ideally we can connect it to other apps. I want to be very clear that this is mainly an **object-role** system. I added functionality to handle system-wide permissions, but this is a different sub-system.

AWX patch to use this - https://github.com/ansible/awx/pull/14735

The only reason this is possible at all is because most of the AWX gnarly permission logic exists on a layer of business logic fairly well-separated from the internals of the RBAC system. It's mainly the internals that this offers (although it tacked on a simple API), for object-roles - roles to things. Those things can be in a shallow hierarchy... which is just organizations right now. Inheritance of permissions from more than 1 parent resource has some basic tests, but isn't expected to be used yet.

Right now this is fully functional in AWX, with only 2-3 known integration test failures due to sacrifices that were knowingly made.

TODO:

 - [x] Add views for this new system. Current blocker is that basic details of the API contract are not be agreed on yet. The AWX branch offers a compatibility layer with the old roles system.
 - [x] Add `content_type` to `RoleDefinition`. This will be used for validation of permissions, because listing some permissions require listing other permissions, and most permissions are only valid for a certain model (or a parent of that model).
 - [x] Fix unit tests in AWX. The compatibility layer currently doesn't do this for us.
 - [x] Reconcile the team & organization `members` relationship that already exists in Gateway compatible with this. The role system, today, handles this by putting users in a pre-created team member role. If the role is the source of truth, then we need to add some ORM shortcuts to more easily access it.
 - ~Very large item - migrate galaxy_ng to this system like the AWX migration. This would have to go along with the introduction of an organization model, without organizations this entire system is useless. Ping @cutwater~
 - [x] ~Add to the AWX migration - split the organization admin role into 2 with 1 dealing with orgs / teams / user management, and the other dealing with AWX objects alone. This role will then be migrated into Gateway itself (not ansible_base). Oh yeah, and also convert Gateway to this system... I don't think it has a permissions system right now.~ Not doing this
 - [x] Can we give this a name? DAB RBAC

AlanCoding committed:
Add an RBAC system as part of DAB app system

Key features are
 - support of resource tree permission inheritance
 - support for grouping users into teams
 - efficient generation of querysets
 - endpoints for RBAC models, permission classes

Added feature for tracking roles with relationships

Added feature to manage singleton permissions

Add util for assigning permissions in migrations
  add util for creating permissions in custom model

Use null content_type to identify system roles

From review
Use shared method for add permission determination
Make the RBAC_checks command more pretty
  give non-zero exit code if issues are identified
Use " instead of ' for block strings
Add model to test parent names not matching model

cutwater commented on ansible_base/management/commands/RBAC_checks.py:
```diff
@@ -0,0 +1,63 @@
+    help = "Runs bug checking sanity checks, gets scale metrics, and reccomendations for Role Based Access Control"
+
+    def handle(self, *args, **options):
+        print(f'Inspecting {RoleDefinition.objects.count()} role definitions')
```
You should use `self.stdout.write` instead of `print` in Django management commands.

cutwater replied:
Minor: I would prefer making queries outside of the f-strings.

cutwater commented on ansible_base/migrations/_managed_definitions.py:
```diff
@@ -0,0 +1,100 @@
+import logging
+
+
+logger = logging.getLogger('awx.ansible_base.migrations._managed_definitions')
```
The `ansible_base` is a top-level package, it doesn't seem to be nested in `awx` package. In the existing codebase loggers do not use `awx` prefix.

cutwater replied:
P.S. is there a reason why you don't use `__name__` special variable? It doesn't look like logger name distinct from module's qualified name.

AlanCoding replied:
Thanks for the comments! I implemented all of them except the `__name__` suggestion, because I don't currently see this being done this django-ansible-base project. I would support that change on a whole-project basis, but for the purposes of this feature I want to keep consistency.

newswangerd commented on ansible_base/models/rbac.py:
```diff
@@ -0,0 +1,389 @@
+    def remove_global_permission(self, actor):
+        return self.give_or_remove_global_permission(actor, giving=False)
+
+    def give_or_remove_global_permission(self, actor, giving=True):
```
If this PR isn't intended to be work with global permissions, what do these functions do?

AlanCoding replied:
I don't want global roles, but I wavered. These methods give global roles. However, this doesn't work unless `ROLE_SINGLETON_USER_RELATIONSHIP` is given a value. This won't create the models to do global permissions, the app has to do that, and define in that setting. Then, if enabled, it goes through a _completely different system_ to check global roles for a permission evaluation, and this isn't tracked in the `RoleEvaluation` table at all. That costs an extra query for permission evaluations if the user global-role relationship is defined, and yet another query if team global-role relationships is defined.

When I was less sure about the plan, I thought we could use global roles to migrate galaxy_ng to this system as-is. Now, that seems extremely unlikely and not helpful. I also just wanted to show how global roles could be added without further complicating the core caching logic.

EDIT: global roles are now considered in evaluations. There are settings to disable them, but they work with the existing assignment models. Plans changed dramatically here since Jan 5. Resolved.

newswangerd commented on ansible_base/models/rbac.py:
```diff
@@ -0,0 +1,389 @@
+    )
+
+    content_type = models.ForeignKey(ContentType, on_delete=models.CASCADE)
+    object_id = models.PositiveIntegerField()
```
This needs to be a text or char field to work with UUIDs as well.

AlanCoding replied:
TIL

https://docs.djangoproject.com/en/5.0/ref/contrib/contenttypes/

> The “object_id” field doesn’t have to be the same type as the primary key fields on the related models, but their primary key values must be coercible to the same type as the “object_id” field by its [get_db_prep_value()](https://docs.djangoproject.com/en/5.0/ref/models/fields/#django.db.models.Field.get_db_prep_value) method.

As these docs continue elaborating, they seem to point in the direction of CharField for our needs. As you say.

newswangerd replied:
The main concern for me is the UUID pks in Hub and EDA

AlanCoding replied:
This has been updated. I will work on adding a docs section for it too. Since the Resource API merged its methods are accepted as canonical to some extent. The _specification_ models now use TextField in the same way as the Resource API. This is important so that coherent system-wide lists can be presented to the user. The _evaluation_ model now maintains different tables for pk and uuid resources.

newswangerd commented on docs/rbac.md:
````diff
@@ -0,0 +1,301 @@
+You can customize that with the following setting.
+
+```
+GATEWAY_ROLE_PRECREATE = {
````
Lets avoid mentioning Gateway in this lib.

newswangerd reviewed (commented):
This looks great. I'd like to merge the resource registry and then update this to integrate with the resource registry system. I think once we do that, I'm happy to merge this PR.

newswangerd commented on ansible_base/rbac/models.py:
```diff
@@ -0,0 +1,397 @@
+    role = models.ForeignKey(
+        ObjectRole, null=False, on_delete=models.CASCADE, related_name='permission_partials', help_text=_("The object role that grants this form of permission")
+    )
+    codename = models.TextField(null=False, help_text=_("The name of the permission, giving the action and the model, from the Django Permission model"))
```
Why use `codename` and not a foreign key to the permission model here?

AlanCoding replied:
The problem we're starting out with is that we need to construct a queryset, and the input we're getting is something like `JobTemplate.accessible_objects(user, 'view_jobtemplate')`, and I added aliasing in so that `JobTemplate.accessible_objects(user, 'view')` works, which is how I predict this is always going to work in practice.

This is little different from AWX today

https://github.com/ansible/awx/blob/f69f600cffa9757f961c222a38a82c58378f6a3a/awx/api/views/mixin.py#L151

which, in practice, does `JobTemplate.accessible_objects(self.request.user, 'read_role')`. Existing practices for that were copied over. So the action is taken in as a string and it's filtered that way. It would make sense to me to strip the model name actually, since that's redundant with `content_type_id`.

newswangerd replied:
Does it make more sense to use a foreign key now that we have a concrete permission model?

I'm worried that codenames could change if someone renames a model.

AlanCoding replied:
Renaming a model is unsolved for `auth.Permission`, documented in https://github.com/ansible/django-ansible-base/issues/227

If you rename a model registered with RBAC, it will change the "evaluation" entries in its post_migrate signal. I'm not worried about that.

newswangerd commented on ansible_base/rbac/models.py:
```diff
@@ -0,0 +1,397 @@
+        """
+        # We only have a content_types exception for multiple content types for polymorphic models
+        # for normal models you should not need it, but AWX unified_ models need it to get by
+        filter_kwargs = dict(role__in=actor.has_roles.all(), codename=codename)
```
Could you do this instead?

```
filter_kwargs = dict(role__in=actor.has_roles.filter(role_definition__permissions__codename=codename))
```

This would make it so you don't have to add "codename" to the RoleEvaluation table. Would this be less efficient?

AlanCoding replied:
There's an explicit index for `["role", "content_type_id", "codename"]`. It's true that all foreign keys are going to be an index by itself.

But the logic won't quite work. The `codename` here is to be the resultant object permission. You could be a have the member role for a team, which lists like 2 permissions, both for the team CT. What we need to know from this table is the fact that the team role gives update permission to inventory 52, for example. That data is written very verbosely in a row here.

newswangerd commented on ansible_base/rbac/models.py:
```diff
@@ -0,0 +1,397 @@
+        return (self.codename, self.content_type_id, self.object_id)
+
+    @staticmethod
+    def accessible_ids(cls, actor, codename, content_types=None):
```
Why is this `actor` instead of `user`? Can you call this for teams as well?

newswangerd replied:
Come to think of it, how do users inherit permissions from teams?

AlanCoding replied:
Yes, you can call it for teams, because it relies on the same name relationship for users/teams when it does `actor.has_roles.all()`. I prefer to avoid using this method in a team-centric way, but AWX has a good handful of corner cases that needed it. I reluctantly added support.

Users are a member of a team by virtue of being give a role that grants membership to that team. That role has to contains the "member_team" permission for this to happen. For every team that has a resource object-role, `RoleEvaluation` entries are duplicated with the `role` field being the team membership role.

To rephrase, the point of `RoleEvaluation` is to duplicate the codename/content_object data for every role that inherits those permissions.

cutwater commented on ansible_base/rbac/caching.py:
```diff
@@ -0,0 +1,190 @@
+            organization_id: [team_id, ...]
+        }
+    """
+    org_team_mapping = {}
```
You can use `defaultdict(list)` instead.

AlanCoding replied:
This applies to several cases where `.setdefault` is used. I went ahead and changed all cases to use defaultdict. Similarly for the type hints, that is added in a few cases.

cutwater commented on ansible_base/rbac/caching.py:
```diff
@@ -0,0 +1,190 @@
+    return parent_team_ids
+
+
+def get_org_team_mapping() -> dict:
```
Would be nice to use more specific type hints (e.g. `dict[int, list[int]]` ) in the code.

cutwater commented on ansible_base/rbac/caching.py:
```diff
@@ -0,0 +1,190 @@
+'''
+
+
+def all_team_parents(team_id: int, team_team_parents: dict, seen=None) -> set:
```
`seen` parameter annotation is missing. Probably `Optional[set]` ?

AlanCoding commented:
I'm still looking at doing work for https://github.com/ansible/django-ansible-base/issues/78, but I don't see it as urgent or blocking of merge. So I'm going to go ahead and say this is out of draft.

Dostonbek1 commented on ansible_base/rbac/api/views.py:
```diff
@@ -0,0 +1,99 @@
+class RoleUserAssignmentViewSet(BaseAssignmentViewSet):
+    """
+    Use this endpoint to give a team permission to a resource or an organization.
+    The needed data is the team, the role definition, and the object id.
```
docstrings seem to be swapped with team, right?

AlanCoding replied:
Yep.

I should provide some context on why a 2-line class has a paragraph docstring. This text will get rendered in the browsable API. So this is directly user/client facing. Probably good to get the model right.

newswangerd commented on ansible_base/rbac/api/serializers.py:
```diff
@@ -0,0 +1,198 @@
+        return perms
+
+    def get_dynamic_object(self, data):
+        codename = data.rsplit('.')[-1]
```
Don't remove the app label from the model name. It can result in this error if multiple apps have a model with the same name: 

```
pulp-1          |   File "/src/django-ansible-base/ansible_base/rbac/api/serializers.py", line 67, in to_internal_value
pulp-1          |     return self.get_dynamic_object(data)
pulp-1          |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pulp-1          |   File "/src/django-ansible-base/ansible_base/rbac/api/serializers.py", line 92, in get_dynamic_object
pulp-1          |     return permission_registry.content_type_model.objects.get(model=model)
pulp-1          |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pulp-1          |   File "/usr/local/lib/python3.11/site-packages/django/db/models/manager.py", line 87, in manager_method
pulp-1          |     return getattr(self.get_queryset(), name)(*args, **kwargs)
pulp-1          |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pulp-1          |   File "/usr/local/lib/python3.11/site-packages/django/db/models/query.py", line 640, in get
pulp-1          |     raise self.model.MultipleObjectsReturned(
```

jbradberry replied:
Endorsed.

AlanCoding replied:
The `data` isn't giving a Django permission reference in the format `"{app_name}.{codename}"` like you are thinking. The DAB resource registry app introduced model references along the lines of `"{service_name}.{model_name}"` like `"awx.inventory"`. This is using permission references according to `"{service_name}.{codename}"`, like `"awx.change_inventory"`, and yes, this requires unique codenames among the usage of this, just like the resource registry does for model names. That's not saying you can't, for example, have proxy models. But you cannot track multiple proxied models in the resource registry, and you can't do it with this RBAC app either. Right now, this only operate for a single service.

I am absolutely not changing this, this is driven by much broader commitments that I have no control over. You can't get an app_name because there is no app_name to obtain. This is a model interchange format that goes between multiple different services. AWX, for instance uses `auth.User` and eda-server uses `core.User`. The interchange format treats these as the same model.

If this is really a problem, see what AWX and eda-server already have with the `DABPermission` model, which uses utils from here to only populate the entries with models tracked in the permission_registry. That works very well. If that can't work with pulp, then the query here can be narrowed to the registered content types.

newswangerd replied:
@AlanCoding Ah, my bad. I misread the code. Do you have a fix for this? I'm guess it' was caused by all the group proxy models we have.

AlanCoding replied:
Yes, I pushed something I believe to be a fix, neglected to update this comment chain. Model `ProxyInventory` was introduced (not actually a proxy model) which abstractly is supposed to cover what you reported here.

This feedback is incredible btw. Test models and tests are introduced as these come in.

newswangerd commented on ansible_base/rbac/api/serializers.py:
```diff
@@ -0,0 +1,198 @@
+    def create(self, validated_data):
+        rd = validated_data['role_definition']
+        model = rd.content_type.model_class()
+        obj = model.objects.get(id=validated_data['object_id'])
```
This breaks when the PK field name is not `id`. Pulp uses `pulp_id` for example.

```
pulp-1          |   File "/src/django-ansible-base/ansible_base/rbac/api/serializers.py", line 154, in create
pulp-1          |     obj = model.objects.get(id=validated_data['object_id'])
pulp-1          |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pulp-1          |   File "/usr/local/lib/python3.11/site-packages/django/db/models/manager.py", line 87, in manager_method
pulp-1          |     return getattr(self.get_queryset(), name)(*args, **kwargs)
pulp-1          |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pulp-1          |   File "/usr/local/lib/python3.11/site-packages/django/db/models/query.py", line 623, in get
pulp-1          |     clone = self._chain() if self.query.combinator else self.filter(*args, **kwargs)
pulp-1          |                                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pulp-1          |   File "/usr/local/lib/python3.11/site-packages/django/db/models/query.py", line 1436, in filter
pulp-1          |     return self._filter_or_exclude(False, args, kwargs)
pulp-1          |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pulp-1          |   File "/usr/local/lib/python3.11/site-packages/django/db/models/query.py", line 1454, in _filter_or_exclude
pulp-1          |     clone._filter_or_exclude_inplace(negate, args, kwargs)
pulp-1          |   File "/usr/local/lib/python3.11/site-packages/django/db/models/query.py", line 1461, in _filter_or_exclude_inplace
pulp-1          |     self._query.add_q(Q(*args, **kwargs))
pulp-1          |   File "/usr/local/lib/python3.11/site-packages/django/db/models/sql/query.py", line 1546, in add_q
pulp-1          |     clause, _ = self._add_q(q_object, self.used_aliases)
pulp-1          |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pulp-1          |   File "/usr/local/lib/python3.11/site-packages/django/db/models/sql/query.py", line 1577, in _add_q
pulp-1          |     child_clause, needed_inner = self.build_filter(
pulp-1          |                                  ^^^^^^^^^^^^^^^^^^
pulp-1          |   File "/usr/local/lib/python3.11/site-packages/django/db/models/sql/query.py", line 1427, in build_filter
pulp-1          |     lookups, parts, reffed_expression = self.solve_lookup_type(arg, summarize)
pulp-1          |                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pulp-1          |   File "/usr/local/lib/python3.11/site-packages/django/db/models/sql/query.py", line 1237, in solve_lookup_type
pulp-1          |     _, field, _, lookup_parts = self.names_to_path(lookup_splitted, self.get_meta())
pulp-1          |                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pulp-1          |   File "/usr/local/lib/python3.11/site-packages/django/db/models/sql/query.py", line 1725, in names_to_path
pulp-1          |     raise FieldError(
pulp-1          | django.core.exceptions.FieldError: Cannot resolve keyword 'id' into field. Choices are: alternatecontentsourcepath, ansible_crossrepositorycollectionversionindex, content, core_pulp_exporter, description, distributions, download_logs, gpgkey, group_roles, last_synced_metadata_time, name, next_version, private, pulp_created, pulp_domain, pulp_domain_id, pulp_id, pulp_labels, pulp_last_updated, pulp_type, pulpimporterrepository, remote, remote_id, repositories, repository_ptr, repository_ptr_id, repositorycontent, retain_repo_versions, uploads, upstream_repositories, user_hidden, user_roles, versions
```

newswangerd replied:
there are a few other places where `content_object.id` is used as well.

AlanCoding replied:
test_app added the `PositionModel` model to hunt down remaining occurrences of this. Thanks.

newswangerd commented on ansible_base/lib/dynamic_config/dynamic_settings.py:
```diff
@@ -107,3 +107,36 @@
+
+if 'ansible_base.rbac' in INSTALLED_APPS:
+    # Settings for the RBAC system, override as necessary in app
+    ANSIBLE_BASE_ROLE_PRECREATE = {
```
Can you document what all of these settings do? Settings like `ANSIBLE_BASE_TEAM_TEAM_ALLOWED` are very confusing.

jbradberry replied:
Endorsed.

AlanCoding replied:
An issue about how to handle the lot of these https://github.com/ansible/django-ansible-base/issues/149

AlanCoding replied:
I pushed a commit that put this together with other `ANSIBLE_BASE_ALLOW_*` settings. That's not hard to follow is it? It allows or disallows stuff. I don't want to allow stuff that we didn't allow in AWX, and that's why it was still on my TODO list here. I closed related issue & Jira. Have a look, if you will. Marking this conversation as resolved.

newswangerd commented on ansible_base/rbac/evaluations.py:
```diff
@@ -0,0 +1,115 @@
+        else:
+            name = codename
+    if name.startswith('add'):
+        if model._meta.model_name != 'organization':
```
This is untenable for hub. A lot of our "add" actions are based on some other parent objects. For example when a user adds a collection to the system, we check to see if they have "add" permissions on the namespace. This can't just be hardcoded to only work with "organization".

We also want to support global add permissions.

AlanCoding replied:
This was sloppy, the "organization" string shouldn't be hard coded anywhere like this. The criteria at this line should be if the model has no child models.

EDIT: Pushed a fix to generalize this and added a number of related assertions in tests.

resolving conversation as fixed.

newswangerd commented:
@AlanCoding I have serious concerns about the way we're using django permissions here. This entire system is built on the notion that a permission code name follows this schema: `{action}_{model_name}`. But that's not true at all. You can create permissions with any codename that you want. Here are some examples:

```
In [14]: content_type = ContentType.objects.get_for_model(AnsibleRepository)

In [15]: permission = Permission.objects.create(
    ...:     codename="can_publish",
    ...:     name="Can Publish Posts",
    ...:     content_type=content_type,
    ...: )

In [16]: permission.codename
Out[16]: 'can_publish'

In [17]: permission = Permission.objects.create(
    ...:     codename="I'm a lovely coconut",
    ...:     name="",
    ...:     content_type=content_type,
    ...: )

In [18]: permission.codename
Out[18]: "I'm a lovely coconut"
```

I can even create permissions for one object type that mimic another object type:

```
In [27]: content_type = ContentType.objects.get_for_model(AnsibleDistribution)

In [28]: permission = Permission.objects.create(
    ...:     codename="add_ansiblerepository",
    ...:     name="",
    ...:     content_type=content_type,
    ...: )

In [29]: permission.codename
Out[29]: 'add_ansiblerepository'
```

Codenames also aren't unique. In this example I created an "I'm a lovely coconut" on the `AnsibleDistrobution` and `AnsibleRepository` models: 

```
In [23]: Permission.objects.filter(codename="I'm a lovely coconut")
Out[23]: <QuerySet [<Permission: ansible | ansible distribution | >, <Permission: ansible | ansible repository | >]>
```

This is also true for models that have the same name. This includes proxy models. Here's an example from hub:

```
In [25]: ContentType.objects.filter(model="group")
Out[25]: <QuerySet [<ContentType: auth | group>, <ContentType: core | group>, <ContentType: galaxy | group>]>

In [26]: Permission.objects.filter(codename="add_group")
Out[26]: <QuerySet [<Permission: auth | group | Can add group>, <Permission: core | group | Can add group>, <Permission: galaxy | group | Can add group>]>
```

I'm not sure where this will bite us, but I'm certain it will. We depend on several third party django apps that we can't control. It's very possible that there will be unintended consequences here, especially around system permissions. If my user has `add_group` permissions, which group does that correspond to?

I'm not sure what the solution is here, but I think we should ship a custom permission model that has more robust schema validation.

As an aside, I'm really not a fan of `_` as delimiter. `_` is super useful for readability and we lose that here. I'd prefer to see something like `{model}:{action}` or `{service_type}:{model}:{action}`

newswangerd commented:
I'm also torn on this ^ because using the django permission codenames will make this WAY easier to implement in Hub. With a custom codename schema we'd need to implement some kind of translation layer.

jbradberry commented on ansible_base/rbac/api/serializers.py:
```diff
@@ -0,0 +1,232 @@
+    """
+
+    default_error_messages = serializers.PrimaryKeyRelatedField.default_error_messages
+    psuedo_field = None  # define in subclass
```
pseudo

jbradberry replied:
Also not used by anything?

AlanCoding replied:
ah you're right, I used it for grabbing the field help_text and for some reason or another it didn't work and something else was done.

jbradberry commented on ansible_base/rbac/api/serializers.py:
```diff
@@ -0,0 +1,232 @@
+        self.html_cutoff_text = kwargs.pop('html_cutoff_text', self.html_cutoff_text)
+
+        self.allow_blank = kwargs.pop('allow_blank', False)
+        super(serializers.ChoiceField, self).__init__(**kwargs)
```
This class is `ChoiceLikeMixin`, not `ChoiceField`.  This probably underscores that we should just use the Python 3 pattern of doing away with the params to `super()`.

AlanCoding replied:
This is, very unfortunately, intended.

Clients need to be able to get available choices for a RoleDefinition `permissions` which is a multi-select type of field. There are 2 ways you could do this. You could have an `/api/v1/pemissions/` endpoint... and at some point maybe this makes sense, but we're bring-your-own-permission model. The other way is to show choices in OPTIONS.

You can't do a many-related field (like a slug related field) and show choices in OPTIONS. Because DRF doesn't let you.

https://github.com/encode/django-rest-framework/blob/a45432b54de88b28bd2e6db31acfe3dbcfc748dd/rest_framework/metadata.py#L142-L151

This hard duck typing of `isinstance(field, (serializers.RelatedField, serializers.ManyRelatedField))` means that we can't show choices in OPTIONS if its of this (most correct) type. That's why I'm using a `ChoiceField` to start with, but the `ChoiceField` is extremely unwilling to allow the choices to be dynamic.

https://github.com/encode/django-rest-framework/blob/a45432b54de88b28bd2e6db31acfe3dbcfc748dd/rest_framework/fields.py#L1394

The problem isn't performance or anything, the problem there was load order. We really don't want to reference stuff from our registry until after apps are ready, and this `__init__` method very much ran before that. This is one reason I created https://github.com/ansible/django-ansible-base/pull/127, because the settings are available much earlier during import than the registry for this or the other app, dab_resource_registry. So to avoid rendering choices before apps are ready, yes, I jumped over the `ChoiceField.__init__` to its parent and set the other things it did manually.

This is absolutely terrible in every way. I do hope we burn it to the ground, eventually, but this does what we want for right now.

jbradberry replied:
Fair enough, I see what you mean.  However, I do prefer DjRF's thinking on this -- the principled thing to do here is have a `permissions` endpoint.

jbradberry replied:
> and this `__init__` method very much ran before that.

This I'm having a hard time with.  `__init__` methods are not generally run at import time, unless you are (directly or indirectly) instantiating the object at the module level.  Are we doing that somewhere?

AlanCoding replied:
Sorry for the response delay. DRF fields are initialized at import time. It is rather annoying.

Related - we are planning to move the Permission model into this library as a concrete model. That will eventually allow for unwinding some of the hacks here.

jbradberry commented on ansible_base/rbac/api/serializers.py:
```diff
@@ -0,0 +1,232 @@
+from ansible_base.rbac.validators import validate_permissions_for_model
+
+
+class ChoiceLikeMixin(serializers.ChoiceField):
```
TBH I'm highly skeptical about re-rolling this functionality.  I strongly suspect that `serializers.RelatedField` (or friends) is going to get you most if not all of what you are trying to do here.

Also, not really a mixin.

jbradberry commented on ansible_base/rbac/api/serializers.py:
```diff
@@ -0,0 +1,232 @@
+    default_error_messages = serializers.PrimaryKeyRelatedField.default_error_messages
+    psuedo_field = None  # define in subclass
+
+    def get_model_for_init(self):
```
As far as I can tell, this doesn't hook into anything, either DjRF or DAB.

jbradberry commented on ansible_base/rbac/api/serializers.py:
```diff
@@ -0,0 +1,232 @@
+        return self._grouped_choices
+
+    @property
+    def choices(self):
```
We should probably use https://docs.djangoproject.com/en/5.0/ref/utils/#django.utils.functional.cached_property for these.

jbradberry replied:
This didn't really accomplish anything, since you are still manually managing an attribute to be the cache.  The point about my comment was that `@cached_property` would remove the need to have `self._choices`.

AlanCoding replied:
I pushed a change to clean that up a little bit, and further lean into reliance on `cached_property`. And I hope I can get it cleaned up a good deal more with the custom permission model later.

jbradberry commented on ansible_base/rbac/permission_registry.py:
```diff
@@ -0,0 +1,155 @@
+
+    @cached_property
+    def content_type_model(self):
+        return self.apps.get_model('contenttypes.ContentType')
```
I'm not sure this makes sense.

- `apps.get_model` is a flat lookup in a dictionary mapping names to classes.  It's already going to be fast, and caching it isn't going to make it any faster.  (https://github.com/django/django/blob/ef2434f8508551fee183079ab471b1dc325c7acb/django/apps/config.py#L235)
- why are we getting it dynamically anyway?  it's not going to be changing to the best of my understanding.
- it feels like a bit of a weird fit to put the `ContentType` as an attribute onto a permissions registry

AlanCoding replied:
I agree, it shouldn't be cached.

ContentType is only like this for future reasons. The permission class is declared as a swappable model right now. That means that a project points a setting to the model to use, and that model has to meet certain minimum criteria. That could be the `auth.Permission` model or a custom model. The reason why I have ContentType handled in a similar way is because I intend to make it swappable for https://github.com/ansible/django-ansible-base/issues/80. We probably _can_ merge the permission models, but the ContentType will need to swap depending on if it's a plain app, or a controller of external permissions.

jbradberry replied:
... swappable `ContentType`?  A key part of the Django internals?

**edit:** I see, non-locality of models.  Reserving judgement.

AlanCoding replied:
You can't durably add a `ContentType` entry for a model that doesn't exist in the local server. That's what I have in mind. The model will exist in a _different_ server, but permissions for those objects will be tracked locally. That's not a problem for the `object_id` of a GenericForeign key because it's not relational in the first place, lol. But it doesn't work with the stock `ContentType` model.

jbradberry commented on ansible_base/rbac/urls.py:
```diff
@@ -0,0 +1,14 @@
+
+app_name = AnsibleRBACConfig.label
+
+api_version_urls = [
```
At first I had no idea what was going on with this, but now that I do I'm unhappy about it.

Why are we doing this?  https://github.com/ansible/django-ansible-base/blob/devel/ansible_base/lib/dynamic_config/dynamic_urls.py

As far as I can tell, the only benefit is that it avoids writing a urls file.  And that is IMO a dubious benefit, since the project can't then opt-out or customize what is going on.

We seem to have 3 different flavors of url that dynamic_urls.py is processing, but as far as I can tell it isn't actually **doing** anything with that.  Why?

As far as I can tell, there's no reason why DAB couldn't either:

a) require projects to create a urls.py file that explicitly integrates only the urls for the apps that it is using, or
b) have DAB expose a top-level urls.py file in the normal Django style that conditionally does an `include()` for each module's urls.py if it is present in the settings

AlanCoding replied:
This is pretty standard from https://github.com/ansible/django-ansible-base/blob/devel/docs/apps_or_lib.md

This RBAC app only adds "version" URLs, because it just adds paths like `/api/v2/role_definitions/`

jbradberry replied:
My complaint is only about 10% about the non-standard module attributes we are using to accumulate the urls, and more like 90% about the weird and brittle machinery in https://github.com/ansible/django-ansible-base/blob/devel/ansible_base/lib/dynamic_config/dynamic_urls.py

AlanCoding commented:
> This entire system is built on the notion that a permission code name follows this schema: `{action}_{model_name}`.

@newswangerd in test_app, model `WeirdPerm` has been added to test violations of this. Those codenames are:
 - `"I'm a lovely coconut"`
 - `"crack"`

The test case I still have TODO is to add a proxy model. Could be same app, different app, or maybe both test cases would be good. Repeating my intent, only 1 proxy model for a concrete model will be allowed to be registered.

Like you say, we have several options for the permission model. We could introduce a `AbstractPermission` abstract model to share, but still let it be swappable. Or we could have this app manage its own permission model, and the AWX and eda-server work to catch up would be very little, because they are both using a new, dedicated permission model. So these are very flexible.

jbradberry commented on test_app/views.py:
```diff
@@ -21,14 +24,40 @@ class TeamViewSet(TestAppViewSet):
     serializer_class = serializers.UserSerializer
 
+    def get_queryset(self):
+        return User.objects.all()
```
Minor: Unless you are doing something special per-user, it makes more sense to set the `queryset` attribute on the class.

AlanCoding replied:
I owe an update on this. The issue here is - `User.objects.all()` is the _wrong answer_. That says "always show all users". But awkwardly, users don't have permissions or roles in this system. In the AWX system, permission to edit users came from _organization_ permissions. But it gets worse.

Due to reasons, there really does need to be canonical rules for who can see users (this queryset, dependent on requesting user) and do the other user CRUD actions. There are legacy rules to consider, and they are not the same. They're not close to being the same. Implementing a set of rules for user CRUD would be a simple task - frankly easy. Getting people to understand that there is a conflict of rules, agree that a single set of rules is needed, and then agree on what those rules should be... is an exhausting thought. If I were to hand-wave that as done, then I would agree that it would be a validly in the domain of DAB RBAC.

jbradberry replied:
Both `.list()` and `.get_object()` on the base `ModelViewSet` class by default call `self.filter_queryset(self.get_queryset())`.  The key for permissions control is either `.filter_queryset` (for both `.list` and `.get_object`, if we want to prevent even seeing some things) or `.check_object_permissions` (specifically for `.get_object`), not in hand-crafting a starting queryset.

- https://www.cdrf.co/3.14/rest_framework.viewsets/ModelViewSet.html#get_object
- https://www.cdrf.co/3.14/rest_framework.viewsets/ModelViewSet.html#list
- https://www.cdrf.co/3.14/rest_framework.viewsets/ModelViewSet.html#filter_queryset

jbradberry replied:
We might be talking a bit at cross-purposes here.  My concern is more making use of more of the customization hooks that Django REST Framework provides, rather than overloading everything into `.get_queryset()`.

AlanCoding replied:
Sure, agreed in principal on moving away from `get_queryset` to `filter_queryset`. That's no big deal. I think we do need to allow the bound method `access_qs` accept a base queryset, which is no technical problem at all.

AlanCoding replied:
I looked into this more, and I just don't want to do it right now. This creates a lot of new disturbances within test_app. Could we maybe file an issue to do this later?

jbradberry commented on ansible_base/rbac/management/commands/RBAC_checks.py:
```diff
@@ -0,0 +1,50 @@
+        indexed_rds = defaultdict(list)
+        for rd in RoleDefinition.objects.prefetch_related('permissions'):
+            perm_list = list(rd.permissions.values_list('codename', flat=True))
+            if 'view' not in ' '.join(perm_list):
```
This seems like not the best idea.  Perhaps `if not any(p.startswith('view_') for p in perm_list):` (or whatever the pattern is for these permission tokens)?

We just had the bug in AWX with substring matching on exceptions, and I know that has to do with language localization which will not be operative here, but this also seems like a bug waiting to happen due to its fuzziness.

jbradberry commented on ansible_base/rbac/management/commands/RBAC_checks.py:
```diff
@@ -0,0 +1,50 @@
+        self.stdout.write('  checking for duplicate role definitions')
+        for perm_set, rd_list in indexed_rds.items():
+            if len(rd_list) > 1:
+                self.stdout.write('Found duplicate role definitions with same permissions list:')
```
Minor: we should probably make use of `self.style.*` for many of these.  This one seems like `self.style.NOTICE` or `self.style.WARNING`.

https://docs.djangoproject.com/en/4.2/howto/custom-management-commands/#django.core.management.BaseCommand.style

jbradberry commented on ansible_base/rbac/management/commands/RBAC_checks.py:
```diff
@@ -0,0 +1,50 @@
+
+
+class Command(BaseCommand):
+    help = "Runs bug checking sanity checks, gets scale metrics, and reccomendations for Role Based Access Control"
```
Typo: recommendations

jbradberry commented on ansible_base/rbac/management/commands/RBAC_checks.py:
```diff
@@ -0,0 +1,50 @@
+from ansible_base.rbac.models import ObjectRole, RoleDefinition
+
+
+class Command(BaseCommand):
```
This command doesn't seem to do anything other than output human-oriented text.  Seems like it ought to fail when it detects something wrong or provide some program-consumable output or something.

Or perhaps this shouldn't even be a management command, and should instead be a system check.  https://docs.djangoproject.com/en/5.0/topics/checks/

AlanCoding replied:
This is too slow for Django checks... way too slow. I always thought this command might outlive its usefulness. I don't want to make the call it delete it until I've gotten more checks done with large data sets, which I've only loaded up in AWX before.

jbradberry replied:
Fair enough, but it IMO should have at least a fail / no-fail return code.  Styling would be nice on top of that.

jbradberry commented on ansible_base/rbac/api/permissions.py:
```diff
@@ -0,0 +1,110 @@
+            queryset = self._queryset(view)
+            model_cls = queryset.model
+            parent_field_name = permission_registry.get_parent_fd_name(model_cls)
+            if parent_field_name not in request.data or parent_field_name is None:
```
Minor: for better short circuiting, these two clauses should be reversed.

jbradberry commented on ansible_base/rbac/api/permissions.py:
```diff
@@ -0,0 +1,110 @@
+            return [f'{special_action}_{model_cls._meta.model_name}']
+        perms = super().get_required_object_permissions(method, model_cls)
+        # Remove add permissions, since they are handled in has_permission
+        return [p for p in perms if 'add_' not in p]
```
`.startswith()`.  This is another substring match potential bug, albeit one less likely to fire off than the first one I pointed out.

AlanCoding replied:
I introduced an entire `is_add_perm` method, so that we collect all the cases where this determination happens.

jbradberry commented on ansible_base/rbac/caching.py:
```diff
@@ -0,0 +1,214 @@
+logger = logging.getLogger('ansible_base.rbac.caching')
+
+
+'''
```
Style: for whatever weird historical reasons the AWX codebase is littered with these single-triples (as well as having double-triples), but for this new codebase we should IMHO standardize and do all double-triples.

fosterseth commented on ansible_base/rbac/models.py:
```diff
@@ -0,0 +1,612 @@
+
+
+class RoleUserAssignment(AssignmentBase):
+    role_definition = models.ForeignKey(
```
why do we need `role_definition` on both RoleUserAssignment and ObjectRole?

AlanCoding replied:
Good question! The core unchangeable aspects are:
 - the `ObjectRole` model maps a `RoleDefinition` and an object via a GenericForeignKey
 - The assignment models are a through model between `ObjectRole` and the actor models of user & team

Everything else on the assignment models is cruft. You can call it "cached" data. It's not the source of truth. So you would say that `ObjectRole` is the source-of-truth about the `role_definition` field. Since an assignment has 1 and exactly 1 object role, there's no ambiguity and the assignment's role definition should be the object role's role definition. In code you see just that, it is literally outright copied.

Why? For the UI. For clients. At a point, the decision was made to not expose `ObjectRole` objects in the API. While we could keep it on the related `object_role`, this would create headache for filtering and serializers and whatnot.

Since the assignment models and the object-role models are all immutable, that makes this exceptionally easy to get away with.

fosterseth replied:
> Why? For the UI. For clients. 

perfect, make sense. thanks!

fosterseth commented on ansible_base/rbac/triggers.py:
```diff
@@ -0,0 +1,336 @@
+        ObjectRole.objects.filter(users__isnull=True, teams__isnull=True).delete()
+
+    ct = ContentType.objects.get_for_model(instance)
+    ObjectRole.objects.filter(content_type=ct, object_id=instance.id).delete()
```
in the case where I have Org > Inventory, and I delete an Inventory, I notice the RoleEvaluations associated with that Inventory (at the Org level) are not deleted right away. Not until `compute_object_role_permissions` is called again.

Trying to think if having stale RoleEvaluation entries could lead to any issues. I guess since the Inventory no longer exists, these RoleEvaluations are benign

AlanCoding replied:
This test case passes for me:

```python
@pytest.mark.django_db
def test_seth(org_inv_rd, rando, organization, inventory):
    org_inv_rd.give_permission(rando, organization)
    inv_ct = permission_registry.content_type_model.objects.get_for_model(inventory)
    assert RoleEvaluation.objects.filter(object_id=inventory.id, content_type_id=inv_ct.id).count() == 2
    inventory.delete()
    assert RoleEvaluation.objects.filter(object_id=inventory.id, content_type_id=inv_ct.id).count() == 0
```

Now if you're talking about the case where `ANSIBLE_BASE_CACHE_PARENT_PERMISSIONS` is set to True, it will make `RoleEvaluation` entries corresponding to the `organization` object too. It should still do that even if there are no inventories in the organization, and rebuilding the cache shouldn't remove those entries either.

fosterseth replied:
hmm interesting that test makes sense, here is what I was seeing in shell_plus
```
In [21]: RoleEvaluation.objects.filter(content_type_id=31).count()
Out[21]: 24

In [22]: i = Inventory.objects.create(name="invA", organization=o)
2024-03-05 18:33:54,836 DEBUG    ansible_base.rbac.caching Adding 4 object-permissions to ObjectRole(pk=1, organization=1)
2024-03-05 18:33:54,836 INFO     ansible_base.rbac.caching Adding 4 object-permission records

In [23]: RoleEvaluation.objects.filter(content_type_id=31).count()
Out[23]: 28
```

so far so good, we added an inventory to an org, which gets 4 more RoleEvaluation entries, one for each permission

```
In [24]: i.delete()
Out[24]: (1, {'test_app.Inventory': 1})

In [25]: RoleEvaluation.objects.filter(content_type_id=31).count()
Out[25]: 28
```

the count didn't go down as I expected

```
In [26]: i = Inventory.objects.create(name="invA", organization=o)
2024-03-05 18:34:12,372 DEBUG    ansible_base.rbac.caching Removing 4 object-permissions from ObjectRole(pk=1, organization=1)
2024-03-05 18:34:12,372 DEBUG    ansible_base.rbac.caching Adding 4 object-permissions to ObjectRole(pk=1, organization=1)
2024-03-05 18:34:12,372 INFO     ansible_base.rbac.caching Adding 4 object-permission records
2024-03-05 18:34:12,390 INFO     ansible_base.rbac.caching Deleting 4 object-permission records

In [27]: RoleEvaluation.objects.filter(content_type_id=31).count()
Out[27]: 28

```
adding another inventory does trigger some log lines that say 4 were removed, and 4 were added

AlanCoding replied:
Gah! Because `inventory.id` was referenced after it was deleted in my test case. Now I'm reproducing, irrelevant of that setting. Yikes, I'm going to have to spend some time chasing this down and buttoning it up. Nice find!

EDIT: fix and tests have been pushed.

jbradberry commented on ansible_base/rbac/management/commands/RBAC_checks.py:
```diff
@@ -0,0 +1,66 @@
+        if not self.has_issues:
+            self.stdout.write(self.style.SUCCESS('No issues were found'))
+        else:
+            raise CommandError('Checks completed, some potential issues were noted')
```
:+1:

jbradberry commented on ansible_base/lib/utils/models.py:
```diff
@@ -27,6 +27,19 @@ def get_type_for_model(model):
+    Although the function is trivial, this standardizes the criteria.
+    """
+    use_codename = codename
+    if codename.count('.') == 1:
```
:moyai: ?  What's with the dot?

AlanCoding replied:
The reason is to cope with the Django format of permission. Here you can see DRF format permissions according to syntax like `'%(app_label)s.change_%(model_name)s'`, but it wasn't their invention. This syntax ultimately comes from Django. Clarifying - that is ultimately `"{app_label}.{codename}"`. As per other conversations here, this library won't support duplicate codenames, which are actually expected to be quite common (like `change_user` because of using a custom user model). While duplicate codenames are expected to _exist_ in auth.Permission, if you register two models with the same permission codename... it's almost certain that you did something wrong. Like, registering 2 user models, that should raise a red flag.

https://github.com/encode/django-rest-framework/blob/09a0c551ca934e486d2765e042a287c15fa85e75/rest_framework/permissions.py#L280-L289

AlanCoding committed:
Fix bugs from review, rebase in evaluations

Fix bug with stale org-obj evaluations from review
Be more consistent with ContentType usage
Fix test failure from rebase due to postgres and flake8

AlanCoding committed:
Add hack to address AWX schema failures

AlanCoding committed:
Support limited assignment methods for teams - for AWX

Avoid database .get for unsaved objects in pre_save
  discovered via eda-server testing

AlanCoding committed:
Add more validation settings

Fix RBAC_checks management command
  when the thing it is doing is disabled by settings

addition:
Add another setting for disabling custom system role creation

jbradberry commented on ansible_base/rbac/validators.py:
```diff
@@ -0,0 +1,186 @@
+
+def codenames_for_cls(cls) -> set[str]:
+    "Helper method that gives the Django permission codenames for a given class"
+    return set([t[0] for t in cls._meta.permissions]) | set(f'{act}_{cls._meta.model_name}' for act in cls._meta.default_permissions)
```
A very minor thing, but both parts of this have newer idioms that can be used

`set(t[0] for t in cls._meta.permissions)`, or better yet `{t[0] for t in cls._meta.permissions}`

and

`{f'{act}_{cls._meta.model_name}' for act in cls._meta.default_permissions}`

AlanCoding replied:
That situation repeats several other times in this file. I did the rest of this file, but you might also change out the syntax in some other files, which I have not done. I don't have an issue with the bracket syntax, I'm just not so refined, stuck in my old country ways.

AlanCoding committed:
Apply filter backends to RBAC views

AlanCoding committed:
Add system for related object checks in an API

Add plus integration docs for this new feature

Add tests for using unauthenticated client with some views

AlanCoding committed:
Raise error for unregistered models and fix corresponding bugs

Assure that user viewset has minimal functionailty with generic permission class

AlanCoding committed:
Review comment about sets and linter fix

AlanCoding committed:
Assignment permission check alternative where model has no change permission

AlanCoding committed:
Convert DAB RBAC to using its own concrete permission model (#3)

* Initial add of permission model

* Add an early exit for permission creation method

AlanCoding committed:
Check same obj permissions for assignment and revoking

AlanCoding committed:
Code style changes from review

Adopt new created and modified fields in CommonModel
  after rebase - supporting changes

john-westcott-iv commented on ansible_base/lib/dynamic_config/dynamic_settings.py:
```diff
@@ -114,3 +114,35 @@
+    ANSIBLE_BASE_ALLOW_SINGLETON_ROLES_API = True
+
+    # User flags that can grant permission before consulting roles
+    ANSIBLE_BASE_BYPASS_SUPERUSER_FLAGS = ['is_superuser']
```
I was thining about this today. If we are doing auditor as a role could we also do a superuser role and just ignore the is_superuser flag or maybe it means "AAP useruser".

AlanCoding replied:
We don't do a superuser role. AWX did, and it synced superuser role membership with the `is_superuser` "user flag". One of my specific objectives was to kill that off.

Not your question, but the setting right below this existed to support auditor, which is migrated to a role, and that needs to die, so I filed https://github.com/ansible/django-ansible-base/issues/220

I think we should still keep `ANSIBLE_BASE_BYPASS_SUPERUSER_FLAGS` because an app _could_ want to use `is_staff`, and who am I to tell them they're wrong?

To your larger point, jump to a world where we manage "remote" permissions. Yes, you could have a "System Admin" role for the app, which is inferior to `is_superuser`, because it gets synced between multiple services. In case it wasn't clear, `is_superuser` will bypass all other checks. On that :+1: I agree strategically.

john-westcott-iv commented on ansible_base/lib/utils/models.py:
```diff
@@ -34,6 +34,19 @@ def get_type_for_model(model):
+    Although the function is trivial, this standardizes the criteria.
+    """
+    use_codename = codename
+    if codename.count('.') == 1:
```
Would a regex be more efficient here? I didn't test it but I'm thinking something like
```
return re.match(r'^([^\.]+\.)?add_.*$', codename)
```

AlanCoding replied:
valid, but not really priority here

john-westcott-iv commented on ansible_base/lib/utils/models.py:
```diff
@@ -34,6 +34,19 @@ def get_type_for_model(model):
     return underscore(opts.object_name)
 
 
+def is_add_perm(codename: str) -> bool:
```
When my brain saw this I kept thinking shouldn't this be `can_add_perm` but I'm not sure there is any better way to name this. Does `is_perm_add_type` make sense?

john-westcott-iv replied:
Should this function be in lib/utils/models or would it be better served in something like rbac/utils?

john-westcott-iv commented on ansible_base/lib/utils/views/django_app_api.py:
```diff
@@ -29,4 +33,6 @@
 
 class AnsibleBaseDjangoAppApiView(parent_view_class):
-    pass
+    # In case an app is using features like authentication, resources, or RBAC
+    # but not using rest_filters app, we need to specify the same filter backends
+    filter_backends = (TypeFilterBackend, FieldLookupBackend, SearchFilter, OrderByBackend)
```
Wasn't this in another PR? Or are we just merging this as part of this?

AlanCoding replied:
https://github.com/ansible/django-ansible-base/pull/212 has re-done this, and we should use that code. It will cause a conflict, and it should win. The UI needed it, but I guess I might be able to revert it for now. I'm a bit torn, can you look at the other PR and give your preferred plan of action?

john-westcott-iv commented on ansible_base/resource_registry/apps.py:
```diff
@@ -41,14 +41,14 @@ def initialize_resources(sender, **kwargs):
-            content = ContentType.objects.get_for_model(resource.model)
+        for key, resource_config in registry.get_resources().items():
+            content = ContentType.objects.get_for_model(resource_config.model)
 
-            if serializer := resource.managed_serializer:
+            if serializer := resource_config.managed_serializer:
                 resource_type = f"shared.{serializer.RESOURCE_TYPE}"
             else:
                 resource_type = f"{registry.api_config.service_type}.{content.model}"
-            defaults = {"externally_managed": resource.externally_managed, "name": resource_type}
+            defaults = {"externally_managed": resource_config.externally_managed, "name": resource_type}
```
This seems somewhat unrelated to RBAC, are we just changing these just because or is there something prompting this?

AlanCoding replied:
In the serializers I had to write loops doing the same things (so we would process the same type names). So changes here are to keep consistency of variable names.

john-westcott-iv commented on ansible_base/rbac/evaluations.py:
```diff
@@ -0,0 +1,91 @@
+        # values_list will make the return type set[str]
+        permission_qs = DABPermission.objects.values_list('codename', flat=True)
+        self._singleton_permissions = RoleDefinition.user_global_permissions(self, permission_qs=permission_qs)
+        bound_singleton_permissions._team_clear_signal = False
```
Is there any kind of race condition in here around the _team_clear_signal? Do we need this method to be thread safe or atomic anything like that?

AlanCoding replied:
Alright, this sucks. I just didn't want to do nothing here. I was _barely_ convinced that I should allow referencing team global permissions in the first place (you should only be evaluating user permissions), but there was some god-awful special case from AWX that drove it. The only time _this_ should matter would be if you were writing scripts to do something. It's not going to be airtight, but I think the case where it would matter is pretty absurdly niche.

john-westcott-iv commented on ansible_base/rbac/caching.py:
```diff
@@ -0,0 +1,214 @@
+from ansible_base.rbac.permission_registry import permission_registry
+from ansible_base.rbac.prefetch import TypesPrefetch
+
+logger = logging.getLogger('ansible_base.rbac.caching')
```
Do we actually cache anything in here? I don't see an import of django.cache or any lru caches or anything. Should this named compute instead of caching? Or does the caching happen somewhere else?

AlanCoding replied:
@fosterseth made the point (I don't disagree) that I should rename the caching stuff to "evaluations". Catch phrase: it fills in the evaluation table(s). Problem is, I'd have to rename existing evaluations module... probably to "public methods" or something. No blockers to rename stuff (this, in particular, is _very_ internal). I just need a more complete plan.

john-westcott-iv commented on ansible_base/rbac/api/permissions.py:
```diff
@@ -0,0 +1,104 @@
+        return has_super_permission(request.user)
+
+
+class AnsibleBaseObjectPermissions(DjangoObjectPermissions):
```
Why dosen't this inherit from the IsSystemAdminOrAuditor?

AlanCoding replied:
Those (and the AWX permission class) inherit from `BasePermission` which is a super class of `DjangoObjectPermissions`. Yes, I am using some methods from `DjangoObjectPermissions`. This is very dear to me, because it has methods to translate HTTP method & model to permissions. As hard as it was to work that in, I am still clinging for dear life to have something as close to standard as I can. That means using the model `_meta` permissions, and accepting the some call patterns from that.

Now that you mentioned it, I filed https://github.com/ansible/django-ansible-base/issues/221 to make it look _more_ like the DRF class. We need to write this on the whiteboard like Bart Simpson in detention "Don't make things custom when there is no need to"

john-westcott-iv commented on docs/apps/rbac.md:
````diff
@@ -0,0 +1,538 @@
+}
+```
+
+Name can be anything string that's not blank. Description can be any string.
````
`anything string` -> `any string`

john-westcott-iv commented on docs/apps/rbac.md:
```diff
@@ -0,0 +1,538 @@
+From any of the assignment lists, the user can select an assignment to revoke.
+Follow the "url" in the serializer from either user or team assignment lists.
+
+DELETE http://127.0.0.1:8000/api/v1/role_user_assignments/1/
```
Can we be consistent like above and who both a role_user_assignments delete along with a role_team_assignments delete?

AlanCoding replied:
I didn't really grok that sentence.

john-westcott-iv commented on docs/apps/rbac.md:
```diff
@@ -0,0 +1,538 @@
+Instead of giving permission to a single inventory, you can use roles to give
+permission to all inventories inside of an organization.
+
+The difference in the above steps are that
+- when creating a custom role definition the `content_type` would be "shared.organization"
+- when creating the assignment, the `object_id` would be the id of an organization
```
Can we give an example here? Are you saying I would first `/api/v1/role_definition` with a payload of:
```json
{
    "permissions": ["view_inventory"],
    "content_type": "shared.organization",
    "name": "View all org inventories",
    "description": "Custom Org Role"
}
```

And then I would do a second call to `/api/v1/role_team_assignments/` (or role_user_assignments) with a payload of:
```json
{
    "role_definition": <id of the role created>,
    "object_id": <id of the org>,
    "team": 2
}
```

And this would know to let team:id = 2 read all inventories in the org?

newswangerd commented on ansible_base/rbac/models.py:
```diff
@@ -0,0 +1,683 @@
+    class Meta:
+        verbose_name = "permission"
+        verbose_name_plural = "permissions"
+        unique_together = [["content_type", "codename"]]
```
Codnames need to either be unique by themselves or this model needs to have some kind of method for creating a unique permission key using the codename and the object_type. These codenames will have to be used to communicate permissions between systems, and we can't have any ambiguity about what permission a particular codename corresponds to.

I would really prefer the latter, with some unique identifier that includes the ansible service type (awx, galaxy, etc), model name and codename (though the AAP service type isn't strictly necessary here). Maybe we have an "action" field and generate the codename field from the content type name and action? `codename = f"{action}_{content_type.model}"` or something like that. Even better if it's formatted like `{content_type.model}.{action}` for readability :)

Another option would be a UUID field, but that's not readable at all.

AlanCoding replied:
reply in mainline discussion / another issue

john-westcott-iv commented on docs/apps/rbac.md:
```diff
@@ -0,0 +1,538 @@
+of information would give a complete summary of who has access.
+By itself, this does not expand the users who are members of teams.
+Assuming that teams-in-teams are enabled, this would not just be the direct
+members of the teams, but a listing of all users who have indirect membership to the team.
```
Do we have any endpoints that will give this to us or is this something that just has to be computed?

AlanCoding replied:
No https://github.com/ansible/django-ansible-base/issues/150, but I want to make the point that _this_ system with these models is actively working to produce the content of the _old_ access list in AWX in its corresponding PR.

AlanCoding replied:
I removed this section, because we have an issue and this is unmaintainable.

john-westcott-iv commented on docs/apps/rbac.md:
````diff
@@ -0,0 +1,538 @@
+Roles are expected to be user-defined in many apps. Example of creating a custom role definition:
+
+```
+from awx.ansible_base.models.rbac import RoleDefinition
````
Should this be `from awx.ansible_base`?

john-westcott-iv commented on docs/apps/rbac.md:
```diff
@@ -0,0 +1,538 @@
+
+NOTE: At various times a "role definition" may be referred to as just a "role".
+
+Roles are expected to be user-defined in many apps. Example of creating a custom role definition:
```
Where should this process be done? In a migration, in apps.py, somewhere else?

AlanCoding replied:
In a migration, as AWX & eda-server both migrate existing roles.

john-westcott-iv commented on docs/apps/rbac.md:
````diff
@@ -0,0 +1,538 @@
+be made into a resource in the RBAC system by registering that resource in the registry.
+
+```
+from awx.ansible_base.utils.permission_registry import permission_registry
````
`awx.ansible_base`

john-westcott-iv commented on docs/apps/rbac.md:
````diff
@@ -0,0 +1,538 @@
+```
+from awx.ansible_base.utils.permission_registry import permission_registry
+
+permission_registry.register(MyModel, parent_field_name='organization')
````
Should this be done in init or just anywhere in the model's file or somewhere else?

AlanCoding replied:
In all 3 concrete examples we have, it was the `models/__init__.py` or just `models.py`. But I don't want to express any particular recommendation for it. The hard requirement would be sometime before apps are ready, but this is hardly a requirement at all.

john-westcott-iv commented on docs/apps/rbac.md:
```diff
@@ -0,0 +1,538 @@
+Assuming `obj` has a related `organization` which was declared by `parent_field_name` when registering,
+you can give and remove permissions to all objects of that type "in" the organization
+(meaning they have that ForeignKey to the organization).
+- `rd.give_permission(user, organization)` - give execute/view permissions to all job templates in that organization
```
Maybe add a link in here reminding that you defined `rd` as pointing to job templates to guide users as to why this allows all JTs in an organization.

john-westcott-iv commented on docs/apps/rbac.md:
```diff
@@ -0,0 +1,538 @@
+The Django `Permission` entry links to the content type for `MyModel`,
+but these permissions can only apply to the parent model of the object,
+or act as system-wide permissions.
+After a user creates an object, you need to give the user creator permissions.
```
Did you mean `editor permissions`?

john-westcott-iv replied:
Oh, it looks like you actually meant creator. So I have permissions on a parent model and then I create a child model and then I have to give myself creator again on the child? It seems like that second step should be an edit or something on the child.

AlanCoding replied:
> It seems like that second step should be an edit or something on the child.

Yes, it is this. It gives `settings.ANSIBLE_BASE_CREATOR_DEFAULTS` permissions. _Presumably_ you include "change" in that list. Wording is a little tricky here, because of course, it has to have a role definition to do this. So it goes through kind of a wacky complicated process. Firstly it checks if you have all the permissions to the object defined in that setting, so if you do, nothing new is assigned. This should happen quite commonly. Next, it looks to see if an existing role definition gives exactly the requested permissions, and if so, it gives you that role. Finally, the awkward case can happen - there's no existing role to give you those permissions. So it creates a new role and gives you that. This is the challenge in supporting the "add" permission, and there is some precedent of this happening in AWX, although AWX didn't have custom roles, leading to these extra steps...

john-westcott-iv reviewed (commented):
I've not combed over every file but wanted to get a start ont he comments. One general question is the layout of the files inside rbac. FIrst there is an api directory but there are some things not in there like caching and evaluations. The other thing is we have many classes in single views/serializers/modles files. I much prefer a directory with seperate files if possible. But those are kind of personal preferences.

If you want to stick with this layout for this PR I am ok with that but maybe we want to have a "standard" layout for apps in DAB in general?

newswangerd commented:
I'm going to go ahead and approve this, though I do think we need to have more conversations around the permissions model.

AlanCoding commented:
> Does it make more sense to use a foreign key now that we have a concrete permission model?

I wanted to do that before today, but workload from other rebasing and other fixes kept me from it. Described in https://github.com/ansible/django-ansible-base/issues/222

AlanCoding committed:
Update docs syntax and other things from review comments

AlanCoding commented:
@john-westcott-iv https://github.com/AlanCoding/django-ansible-base/pull/4 is ready, in that it is passing tests

> If you want to stick with this layout for this PR I am ok with that but maybe we want to have a "standard" layout for apps in DAB in general?

I think it's a pretty clear practice that `models.py` is common, but with multiple categories, beyond a certain scale you can consider a `models/` folder. That PR does that. I'm not super convinced that it is better, but I'm okay with it. I can push it here before merge if people want that.
